diff -u -r1.31 -r1.31.2.1
--- rdiff-backup/rdiff_backup/restore.py	2003/06/26 07:23:04	1.31
+++ rdiff-backup/rdiff_backup/restore.py	2003/07/23 09:04:09	1.31.2.1
@@ -178,8 +178,7 @@
 		rorp = rf.get_attribs()
 		yield rorp
 		if rorp.isdir():
-			for sub_rf in rf.yield_sub_rfs():
-				for rorp in yield_attribs(sub_rf): yield rorp
+			for sub_rf in rf.yield_sub_rfs(): yield sub_rf.get_attribs()
 
 	def subtract_indicies(cls, index, rorp_iter):
 		"""Subtract index from index of each rorp in rorp_iter

