diff -urN rdiff-backup-0.12.2.org/rdiff_backup/rpath.py rdiff-backup-0.12.2/rdiff_backup/rpath.py
--- rdiff-backup-0.12.2.org/rdiff_backup/rpath.py	2003-08-03 20:39:06.000000000 +0200
+++ rdiff-backup-0.12.2/rdiff_backup/rpath.py	2003-08-04 22:41:30.000000000 +0200
@@ -897,11 +897,15 @@
 		This can be useful for directories.
 
 		"""
-		if not fp:
-			fp = self.open("rb")
-			os.fsync(fp.fileno())
-			assert not fp.close()
-		else: os.fsync(fp.fileno())
+                if not fp: self.conn.rpath.RPath.fsync_local(self)
+                else: os.fsync(fp.fileno())
+
+        def fsync_local(self):
+                """fsync current file, run locally"""
+                assert self.conn is Globals.local_connection
+                fd = os.open(self.path, os.O_RDONLY)
+                os.fsync(fd)
+                os.close(fd)
 
 	def fsync_with_dir(self, fp = None):
 		"""fsync self and directory self is under"""
