--- rdiff-backup-0.12.1/cmodule.c.org	2003-07-22 09:39:06.000000000 +0000
+++ rdiff-backup-0.12.1/cmodule.c	2003-07-22 20:09:57.000000000 +0000
@@ -66,8 +66,8 @@
 	}
   }
 #ifdef HAVE_LARGEFILE_SUPPORT
-  size = PyLong_FromLongLong((LONG_LONG)sbuf.st_size);
-  inode = PyLong_FromLongLong((LONG_LONG)sbuf.st_ino);
+  size = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_size);
+  inode = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_ino);
 #else
   size = PyInt_FromLong(sbuf.st_size);
   inode = PyInt_FromLong((long)sbuf.st_ino);
@@ -75,13 +75,13 @@
   mode = (long)sbuf.st_mode;
   perms = mode & 07777;
 #if defined(HAVE_LONG_LONG) && !defined(MS_WINDOWS)
-  devloc = PyLong_FromLongLong((LONG_LONG)sbuf.st_dev);
+  devloc = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_dev);
 #else
   devloc = PyInt_FromLong((long)sbuf.st_dev);
 #endif
 #if SIZEOF_TIME_T > SIZEOF_LONG
-  mtime = PyLong_FromLongLong((LONG_LONG)sbuf.st_mtime);
-  atime = PyLong_FromLongLong((LONG_LONG)sbuf.st_atime);
+  mtime = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_mtime);
+  atime = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_atime);
 #else
   mtime = PyInt_FromLong((long)sbuf.st_mtime);
   atime = PyInt_FromLong((long)sbuf.st_atime);
@@ -129,7 +129,7 @@
 	/* Device files */
 	char devtype[2];
 #if defined(HAVE_LONG_LONG) && !defined(MS_WINDOWS)
-	LONG_LONG devnums = (LONG_LONG)sbuf.st_rdev;
+	PY_LONG_LONG devnums = (PY_LONG_LONG)sbuf.st_rdev;
 	PyObject *major_num = PyLong_FromLongLong(major(devnums));
 #else
 	long int devnums = (long)sbuf.st_dev;
